common/persistence/visibility/visibility_manager_impl.go

```diff
@@ -141,6 +143,90 @@ func (p *visibilityManagerImpl) ListWorkflowExecutions(
 	return p.convertInternalListResponse(response)
 }
 
+func (p *visibilityManagerImpl) ListChasmExecutions(
+	ctx context.Context,
+	request *manager.ListChasmExecutionsRequest,
+) (*manager.ListChasmExecutionsResponse, error) {
+	response, err := p.store.ListChasmExecutions(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+	executions := make([]*manager.ChasmExecutionInfo, 0, len(response.Executions))
+	for _, exec := range response.Executions {
+		combinedMemo, err := deserializeMemo(exec.Memo)
+		if err != nil {
+			return nil, err
+		}
+
+		var userMemo *commonpb.Memo
+		var chasmMemoPayload *commonpb.Payload
+
+		// Check if archetype matches to decide how to split memo
+		requestedArchetypeIDStr := strconv.FormatUint(uint64(request.ArchetypeId), 10)
+		archetypeMatches := false
+		if archetypePayload, ok := exec.SearchAttributes.GetIndexedFields()[defs.TemporalNamespaceDivision]; ok {
+			var archetypeIDStr string
+			if err := payload.Decode(archetypePayload, &archetypeIDStr); err == nil {
+				archetypeMatches = (archetypeIDStr == requestedArchetypeIDStr)
+			}
+		}
+
+		if archetypeMatches && combinedMemo != nil {
+			// Archetype matches - split memo into user and chasm parts
+			if userPayload, ok := combinedMemo.Fields["__user__"]; ok {
+				var userMemoProto commonpb.Memo
+				if err := payload.Decode(userPayload, &userMemoProto); err == nil {
```

Comment:
Check if `payload.Decode` decodes nil values correctly. If so, then there's no need to do the check in L176